lib/sass/script/functions.rb

```diff
@@ -2078,17 +2092,51 @@ def map_get(map, key)
     # @example
     #   map-merge(("foo": 1), ("bar": 2)) => ("foo": 1, "bar": 2)
     #   map-merge(("foo": 1, "bar": 2), ("bar": 3)) => ("foo": 1, "bar": 3)
+    #   map-merge(("foo": 1, "bar": 2), bar, 3) => ("foo": 1, "bar": 3)
+    #   map-merge((a: (b: c)), a, b, d) => (a: (b: d))
     # @overload map_merge($map1, $map2)
     #   @param $map1 [Sass::Script::Value::Map]
     #   @param $map2 [Sass::Script::Value::Map]
     # @return [Sass::Script::Value::Map]
     # @raise [ArgumentError] if either parameter is not a map
-    def map_merge(map1, map2)
+    # @overload map_merge($map1, $keys..., $map2)
+    #   @param $map1 [Sass::Script::Value::Map]
+    #   @param $keys [[Sass::Script::Value]]
+    #   @param $map2 [Sass::Script::Value::Map]
+    # @return [Sass::Script::Value::Map]
+    # @raise [ArgumentError] if either parameter is not a map
+    def map_merge(map1, *keys)
       assert_type map1, :Map, :map1
+      map2 = keys.pop
+      if map1.is_a? Sass::Script::Value::List
+        map1 = map(map1.to_h)
+      end
```

Comment:
Diving back in to implement this, and have smacked back into the problem that I had the first go around with an iterative approach... which is the same reason that your above code doesn't work. The line where you modify the "current_map[keys[-2]]" doesn't work because you have created new frozen hashes all along the process. Which, we can counter for. But, the fact we are dup-ing like this, means that we throw away all branches of the tree that we don't visit. This walks in like a laser, and nukes everything *not* in it's path. And, the frozen hashes make walking-and-duplicating the tree pretty tricky, since there isn't an easy way to copy the tree.

Hence, my going with recursive, which got around all those issues with pretty minimal code.

I'll keep seeing if I can figure out a way to duplicate-and-modify the frozen tree.